### DISCUSSION TITLE
Dynamically config PartyInfo/From[0]/@type ,PartyInfo/To[0]/@type and Agreement

### DISCUSSION THREAD

**Discussion by TMain-Tomi** [phax/phase4#145] [Q&A]
Dear @phax:

I sent a msg to phase4 without PartyInfo/From[0]/@type and meet the error as follow:

```
2023-07-21T16:28:00.113+08:00 ERROR 13276 --- [nio-9090-exec-1] c.h.phase4.servlet.AS4UnifiedResponse    : HTTP 400: Bad Request: Error validating incoming AS4 message with the profile Peppol
 Following errors are present: [PartyInfo/From[0]/@type must be 'urn:fdc:peppol.eu:2017:identifiers:ap' instead of 'null']
```
If I want to ignore this verification or config a different type (PartyInfo/From[0]/@type, PartyInfo/To[0]/@type and Agreement), may you give me some advices? Thanks.

Best Regards,
Tomi Lin

**Comment by phax** [comment]
@TMain-Tomi can you please explain the bigger context in which you want to use phase4?
I am pretty sure you don't want to use it for Peppol or so.
But it seems like you setup your service on the Peppol components, so you also inherited all the Peppol constraints.
To filter out these constraints, you have to create your own "profile" (like the phase4-peppol-profile) that contains all the constraints and default values in the PMode.

hth, Philip

**Reply by TMain-Tomi** [reply]
Dear @phax,
I very sorry missed your reply.
I base on phase4-spring-boot-demo to realize AS4.
I code AS4CustomIncomingMessageProcessorSPI to implement IAS4ServletMessageProcessorSPI to handle incoming message.
For customer's AS4 server, have to config the fixed value of **party-type and agreement**, like as following:
```
<?xml version="1.0" encoding="UTF-8"?>
<PMode xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns="http://holodeck-b2b.org/schemas/2014/10/pmode" xsi:schemaLocation="http://holodeck-b2b.org/schemas/2014/10/pmode ../../repository/xsd/pmode.xsd">
    <id>user2-to-bifrost</id>
    <mep>http://docs.oasis-open.org/ebxml-msg/ebms/v3.0/ns/core/200704/oneWay</mep>
    <mepBinding>http://docs.oasis-open.org/ebxml-msg/ebms/v3.0/ns/core/200704/push</mepBinding>
    <Initiator>
        <PartyId type="urn:fdc:peppol.eu:2017:identifiers:ap">user2</PartyId>
        <Role>Sender</Role>
        <SecurityConfiguration>
            <Signing>
                <KeystoreAlias password="Password1">user2</KeystoreAlias>
                <Algorithm>http://www.w3.org/2001/04/xmldsig-more#rsa-sha256</Algorithm>
            </Signing>
       </SecurityConfiguration>
    </Initiator>
    <Responder>
        <PartyId type="urn:fdc:peppol.eu:2017:identifiers:ap">Test</PartyId>
        <Role>Receiver</Role>
        <SecurityConfiguration>
            <Encryption>
                <KeystoreAlias>Test1</KeystoreAlias>
                <Algorithm>http://www.w3.org/2001/04/xmlenc#aes256-cbc</Algorithm>
            </Encryption>
        </SecurityConfiguration>
    </Responder>
    <Agreement>
        <name>urn:fdc:peppol.eu:2017:agreements:tia:ap_provider</name>
    </Agreement>
</PMode>
```
Now I want the party type and agreement to be dynamically configured，may you give me some advices?

Best Regards,
Tomi Lin

**Reply by phax** [reply]
Hi Tomi Lin,
PModes are designed to be point-2-point connection parameters, and they don't fit well in dynamic environments.
If you want party type and agreement dynamically changed, and you are manually filling a `AS4ClientUserMessage`, then you simply need to call `.setAgreementRefValue(...)` and `.setFromPartyIDType(...)` and `.setToPartyIDType(...)` AFTER an eventually call to `setPMode`.
hth
